File: Data Structures & Algorithms/duplicate-integer/submission-10.go
func hasDuplicate(nums []int) bool {
    seenBefore := make(map[int]bool, 0)
    for _,num := range nums{
        if _,ok := seenBefore[num]; ok{
           return true 
        }
        seenBefore[num] = true
    }
    return false
}

File: Data Structures & Algorithms/jump-game/submission-10.py
class Solution:
    #1dp dp top down
    def canJump(self, nums: List[int]) -> bool:

        cache = {}
        def canReachEnd(index):
            #at last index -> return true
            if index == len(nums) - 1:
               return True
            #already checked if reahable
            if index in cache:
               return cache[index] 

            #otherwise check if reachable if you make a jump of size 1 to size nums[index]: 
            maxReachableIndex = min(len(nums) - 1, index + nums[index])
            for newIndex in range(index + 1, maxReachableIndex + 1):
                if canReachEnd(newIndex):
                   return True
            
            #else, we can't jum pto end at this end
            cache[index] = False
            return cache[index]

        return canReachEnd(0)
        